Readme.md
# PlayStation 2 HDD checker

This is a fork of original project: https://sites.google.com/view/ysai187/home/projects/hddchecker

## Introduction

HDDChecker is a basic disk diagnostic tool meant for testing the health of your PlayStation 2 console's Harddisk Drive unit.

It was conceived and constructed because I didn't want to see any more poor SCPH-20400 units being cut open, just to have the disks within them taken out for testing. >_>

It'll also be useful for those people who need to check the condition of their HDD units, but don't have a working IDE port on their computers to connect the disk to.

### Features

1. Detects and lists the model, serial, firmware, and S.M.A.R.T. status of HDD unit 0 (Primary Master).
2. Performs a surface scan of the disk.
3. Supports 48-bit LBA disks up to 2TB.
4. Performs zero-filling of the disk with large blocks.
5. With the surface scan, bad sectors found might be remappable[^1].
6. Checks for damage to the APA scheme and PFS partitions.
7. Optimizes the partitions on the HDD to reclaim space.

[^1]: Writing to a to-be-remapped sector (those hard-to-read sectors, as recorded by the disk) may cause them to be remapped. However, this might not be the case for all disks.

### Logging Feature

As of HDDChecker v0.96, a new logging feature has been added.

4 log files will be generated:

- When scanning is done:
  - hdck.log
  - fsck.log
- When optimization is done:
  - fssk.log
  - hdsk.log

No log files will be generated for surface scanning or zero-filling, as the purpose of these log files is to allow what was done during the scanning and optimization operations to be explained.

If you experience problems with HDDChecker's scanning and optimization functionality, please contact the developer and provide the log files.

## Notes/known limitations/bugs

- Disks up to 2 TB are supported.
- Do not use (usually old) disks that are not compliant with the ATA-4 specification. Like with every other PlayStation 2 software out there that supports the ATA interface, the disk is expected to support UDMA mode 4 and S.M.A.R.T.
- If the disk's S.M.A.R.T. status is indicated to be no good (NG status), the disk is about to fail and should be replaced.
- This tool may not be able to successfully remap sectors on all drives, as there isn't an official specification on remapping sectors within the ATA specification. If writing to a bad sector isn't sufficient to cause the disk to automatically remap it, the manufacturer's tools may have to be used instead.
- If a bad sector cannot be remapped, the disk is dying and should be replaced.
- As zeros will be written to the bad sector in an attempt to cause the disk to remap it, the data stored within the bad sector will be lost. Do not choose to remap the sector if it contains data that cannot be lost.

## Supported languages

For more information on supported languages and how support for languages can be completed, click here. A template for translating this software is provided in the downloads section of this page.

Supported languages and their translation status

- Japanese - Oga
- English - Completed and built-in.
- French - ShaolinAssassin
- Spanish - ElPatas
- German - LopoTRI
- Italian - Vash The Stampede
- Dutch - Unassigned
- Portuguese - Gledson999

## Changelog for v0.964 (as of 2019/04/13)

- Fixed glitch when rendering fonts.
- The PS2 can now be shut down when the software is idle, by pressing the power button once. This will also shut down HDDs (USB & PS2 HDD, to avoid causing some devices to emergency park.
- (NEW 2019/01/14) Added Japanese translation. Special thanks to Oga.
- (NEW 2019/01/14) Updated USBHDFSD to fix the bug causing files to be unable to be replaced properly.
- (NEW 2019/04/13) Replaced Japanese font with NotoSansJP-Bold, to reduce space wastage.


CHANGELOG.txt
Changelog
Update 2019/04/13: v0.964 re-released!

Update 2019/01/14: v0.964 re-released!

Update 2019/01/11: v0.964 released!

Update 2018/12/08: v0.963 released!

Update 2018/10/27: v0.962 re-released!

Update 2018/10/21: v0.962 released!

Update 2018/06/20: v0.961 re-released!

Update 2018/06/19: v0.961 re-released!

Update 2018/06/18: v0.961 released!

Update 2018/06/12: v0.96 released!

Update 2018/02/04: v0.95 released!

Update 2017/02/28: v0.942 released!

Update 2016/12/16: v0.941 released!

Update 2016/09/16: v0.94 released!

Update 2016/08/09: v0.932 released!

Update 2016/06/10: v0.931 released!

Update 2016/01/09: v0.93 re-released!

Update 2016/01/06: v0.93 re-released!

Update 2016/01/04: v0.93 released!

Update 2014/06/26: v0.92 re-released!

Update 2014/06/23: v0.92 released!

Update 2013/09/30: Bug report.

Update 2013/08/06: Initial public release.


hdst.c
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <kernel.h>
#include <limits.h>
#include <fileXio_rpc.h>
#include <atahw.h>
#include <hdd-ioctl.h>
#include <sys/fcntl.h>

#include "hdst.h"

int SetHDSTIOBufferSize(int sectors)
{
    return fileXioDevctl("hdst:", HDST_DEVCTL_SET_IO_BUFFER_SIZE, &sectors, sizeof(sectors), NULL, 0);
}

struct AtaDeviceData
{
    u16 IdentificationData[256];
    ata_devinfo_t PS2AtadData;
    int SMARTStatus;
    u32 NumSectors;
    char model[42], serial[22], FWVersion[10];
};
static struct AtaDeviceData AtadDeviceData[NUM_SUPPORTED_DEVICES];

static void TrimWhitespacing(char *string, int maxlen)
{
    int i, CharsStart;
    // Locate the start of the string;
    for (i = 0; i < maxlen; i++) {
        if (string[i] != ' ') {
            CharsStart = i;
            for (i = 0; i < maxlen; i++) {
                if ((string[i] = string[CharsStart + i]) == '\0')
                    break;
            }
            break;
        }
    }
}

int InitATADevice(int unit)
{
    int result, i;
    char DeviceName[8];

    memset(&AtadDeviceData[unit], 0, sizeof(AtadDeviceData[unit]));
    sprintf(DeviceName, "hdst%u:", unit);
    if ((result = fileXioDevctl(DeviceName, HDST_DEVCTL_DEVICE_PS2_DETECT, NULL, 0, &AtadDeviceData[unit].PS2AtadData, sizeof(AtadDeviceData[unit].PS2AtadData))) == 0) {
        if (AtadDeviceData[unit].PS2AtadData.exists) {
            if ((result = fileXioDevctl(DeviceName, HDST_DEVCTL_DEVICE_IDENTIFY, NULL, 0, AtadDeviceData[unit].IdentificationData, sizeof(AtadDeviceData[unit].IdentificationData))) == 0) {
                AtadDeviceData[unit].SMARTStatus = fileXioDevctl(DeviceName, HDST_DEVCTL_DEVICE_SMART_STATUS, NULL, 0, NULL, 0);

                if (!IsPSX() && (AtadDeviceData[unit].IdentificationData[ATA_ID_COMMAND_SETS_SUPPORTED] & 0x400)) {
                    if (AtadDeviceData[unit].IdentificationData[ATA_ID_48BIT_SECTOTAL_HI] == 0)
                        AtadDeviceData[unit].NumSectors = AtadDeviceData[unit].IdentificationData[ATA_ID_48BIT_SECTOTAL_LO] | ((u32)AtadDeviceData[unit].IdentificationData[ATA_ID_48BIT_SECTOTAL_MI] << 16);
                    else // If the disk is larger than 2TB, limit the capacity to 2TB.
                        AtadDeviceData[unit].NumSectors = 0xFFFFFFFF;
                } else {
                    AtadDeviceData[unit].NumSectors = AtadDeviceData[unit].IdentificationData[ATA_ID_SECTOTAL_LO] | ((u32)AtadDeviceData[unit].IdentificationData[ATA_ID_SECTOTAL_HI] << 16);
                }

                for (i = 0; i < 20; i++)
                    ((u16 *)AtadDeviceData[unit].model)[i] = BSWAP16(AtadDeviceData[unit].IdentificationData[27 + i]);
                TrimWhitespacing(AtadDeviceData[unit].model, sizeof(AtadDeviceData[unit].model) - 1);
                for (i = 0; i < 10; i++)
                    ((u16 *)AtadDeviceData[unit].serial)[i] = BSWAP16(AtadDeviceData[unit].IdentificationData[10 + i]);
                TrimWhitespacing(AtadDeviceData[unit].serial, sizeof(AtadDeviceData[unit].serial) - 1);
                for (i = 0; i < 4; i++)
                    ((u16 *)AtadDeviceData[unit].FWVersion)[i] = BSWAP16(AtadDeviceData[unit].IdentificationData[23 + i]);
                TrimWhitespacing(AtadDeviceData[unit].FWVersion, sizeof(AtadDeviceData[unit].FWVersion) - 1);
            }
        } else
            result = -ENODEV;
    }

    return result;
}

int PatchSector(const char *device, u32 lba)
{
    HdstSectorIOParams_t SectorIOParams;

    SectorIOParams.lba     = lba;
    SectorIOParams.sectors = 1;
    return fileXioDevctl(device, HDST_DEVCTL_DEVICE_ERASE_SECTORS, &SectorIOParams, sizeof(SectorIOParams), NULL, 0);
}

int IsATADeviceInstalled(int unit)
{
    return AtadDeviceData[unit].PS2AtadData.exists;
}

const char *GetATADeviceModel(int unit)
{
    return AtadDeviceData[unit].model;
}

const char *GetATADeviceSerial(int unit)
{
    return AtadDeviceData[unit].serial;
}

const char *GetATADeviceFWVersion(int unit)
{
    return AtadDeviceData[unit].FWVersion;
}

u32 GetATADeviceCapacity(int unit)
{
    return AtadDeviceData[unit].NumSectors;
}

int GetATADeviceSMARTStatus(int unit)
{
    return AtadDeviceData[unit].SMARTStatus;
}

#if 0
static u16 DeviceIdentificationInfo[NUM_SUPPORTED_DEVICES][256] ALIGNED(16);

struct AtaDeviceData
{
    int exists;
    u64 NumSectors;
};
static ata_devinfo_t PS2AtaDeviceData[NUM_SUPPORTED_DEVICES];
static struct AtaDeviceData AtadDeviceData[NUM_SUPPORTED_DEVICES];

static const char *xfer_types[] = {
    "PIO",
    "MDMA",
    "UDMA"};

void ShowHDDInfo(int unit)
{
    unsigned int i;
    const char *selected_xfer_type;
    unsigned char selected_xfer_mode;
    char name[41], DeviceName[8];

    printf("HDD unit %u:\n", unit);
    sprintf(DeviceName, "hdst%u:", unit);
    if (unit != 0 || fileXioDevctl(DeviceName, HDST_DEVCTL_DEVICE_IDENTIFY, NULL, 0, DeviceIdentificationInfo[unit], sizeof(DeviceIdentificationInfo[unit])) != 0) {
        printf("\tNot installed.\n");
        AtadDeviceData[unit].exists = 0;
        return;
    }
    AtadDeviceData[unit].exists = 1;

    name[40] = '\0';
    for (i = 0; i < 20; i++)
        ((u16 *)name)[i] = BSWAP16(DeviceIdentificationInfo[unit][27 + i]);
    printf("\tModel:\t%s\n", name);
    name[20] = '\0';
    for (i = 0; i < 10; i++)
        ((u16 *)name)[i] = BSWAP16(DeviceIdentificationInfo[unit][10 + i]);
    printf("\tSerial:\t%s\n", name);
    name[8] = '\0';
    for (i = 0; i < 4; i++)
        ((u16 *)name)[i] = BSWAP16(DeviceIdentificationInfo[unit][23 + i]);
    printf("\tFW Ver:\t%s\n", name);

    printf("\tThis drive supports ");
    if (DeviceIdentificationInfo[unit][ATA_ID_COMMAND_SETS_SUPPORTED] & 0x400) {
        printf("48-bit");
        AtadDeviceData[unit].NumSectors = DeviceIdentificationInfo[unit][ATA_ID_48BIT_SECTOTAL_LO] | ((u32)DeviceIdentificationInfo[unit][ATA_ID_48BIT_SECTOTAL_MI] << 16) | ((u64)DeviceIdentificationInfo[unit][ATA_ID_48BIT_SECTOTAL_HI] << 32);
    } else {
        printf("28-bit");
        AtadDeviceData[unit].NumSectors = DeviceIdentificationInfo[unit][ATA_ID_SECTOTAL_LO] | ((u32)DeviceIdentificationInfo[unit][ATA_ID_SECTOTAL_HI] << 16);
    }
    printf(" LBA addressing.\n");
    printf("\tNumber of sectors: %lu\n", AtadDeviceData[unit].NumSectors);

    /* The default transfer mode is PIO. */
    selected_xfer_type = xfer_types[0];
    selected_xfer_mode = 0;

    printf("\tSupported DMA tranfer modes:\n");
    printf("\t\tMDMA: ");
    if ((DeviceIdentificationInfo[unit][63] & 0x7) & 1)
        printf("0 ");
    if ((DeviceIdentificationInfo[unit][63] & 0x7) & 2)
        printf("1 ");
    if ((DeviceIdentificationInfo[unit][63] & 0x7) & 4)
        printf("2 ");

    /* If any of the MDMA modes were selected: */
    if ((DeviceIdentificationInfo[unit][63] >> 8) & 0x7) {
        selected_xfer_type = xfer_types[1];

        if (((DeviceIdentificationInfo[unit][63] >> 8) & 0x7) & 1)
            selected_xfer_mode = 0;
        else if (((DeviceIdentificationInfo[unit][63] >> 8) & 0x7) & 2)
            selected_xfer_mode = 1;
        else if (((DeviceIdentificationInfo[unit][63] >> 8) & 0x7) & 4)
            selected_xfer_mode = 2;
    }
    printf("\n");

    printf("\t\tUDMA: ");
    if ((DeviceIdentificationInfo[unit][53] & 2) != 0) {
        if ((DeviceIdentificationInfo[unit][88] & 0x3F) & 0x01)
            printf("0 ");
        if ((DeviceIdentificationInfo[unit][88] & 0x3F) & 0x02)
            printf("1 ");
        if ((DeviceIdentificationInfo[unit][88] & 0x3F) & 0x04)
            printf("2 ");
        if ((DeviceIdentificationInfo[unit][88] & 0x3F) & 0x08)
            printf("3 ");
        if ((DeviceIdentificationInfo[unit][88] & 0x3F) & 0x10)
            printf("4 ");
        if ((DeviceIdentificationInfo[unit][88] & 0x3F) & 0x20)
            printf("5 ");
        printf("\n");

        /* If any of the UDMA modes were selected: */
        if ((DeviceIdentificationInfo[unit][88] >> 8) & 0x3F) {
            selected_xfer_type = xfer_types[2];

            if (((DeviceIdentificationInfo[unit][88] >> 8) & 0x3F) & 0x01)
                selected_xfer_mode = 0;
            else if (((DeviceIdentificationInfo[unit][88] >> 8) & 0x3F) & 0x02)
                selected_xfer_mode = 1;
            else if (((DeviceIdentificationInfo[unit][88] >> 8) & 0x3F) & 0x04)
                selected_xfer_mode = 2;
            else if (((DeviceIdentificationInfo[unit][88] >> 8) & 0x3F) & 0x08)
                selected_xfer_mode = 3;
            else if (((DeviceIdentificationInfo[unit][88] >> 8) & 0x3F) & 0x10)
                selected_xfer_mode = 4;
            else if (((DeviceIdentificationInfo[unit][88] >> 8) & 0x3F) & 0x20)
                selected_xfer_mode = 5;
        }
    } else
        printf("Not available.\n");

    printf("\tCurrent transfer mode: %s mode %u\n", selected_xfer_type, selected_xfer_mode);

    if (DeviceIdentificationInfo[unit][82] == DeviceIdentificationInfo[unit][83] == 0x0000 || DeviceIdentificationInfo[unit][82] == DeviceIdentificationInfo[unit][83] == 0xFFFF) {
        printf("\tCommand set notification is NOT enabled or NOT supported.\n");
    } else {
        printf("\tSMART is ");
        if (!(DeviceIdentificationInfo[unit][82] & 1)) {
            printf("NOT ");
        } else
            printf("supported.\n");
    }

    if (DeviceIdentificationInfo[unit][85] == DeviceIdentificationInfo[unit][86] == DeviceIdentificationInfo[unit][87] == 0x0000 || DeviceIdentificationInfo[unit][85] == DeviceIdentificationInfo[unit][86] == DeviceIdentificationInfo[unit][87] == 0xFFFF) {
        printf("\tCommand set enabled notification is NOT enabled or NOT supported.\n");
    } else {
        printf("\tSMART is ");
        if (DeviceIdentificationInfo[unit][85] & 1) {
            printf("ENABLED.\n");
        } else
            printf("DISABLED.\n");
    }


    printf("\tThis drive ");
    if (DeviceIdentificationInfo[unit][ATA_ID_SECURITY_STATUS] & 1) {
        printf("supports");
    } else
        printf("does NOT support");
    printf(" security features.\n");

    printf("\tSecurity features are ");
    if (DeviceIdentificationInfo[unit][ATA_ID_SECURITY_STATUS] & ATA_F_SEC_ENABLED) {
        printf("enabled\n");
    } else
        printf("disabled\n");

    printf("\tDrive is ");
    if (!(DeviceIdentificationInfo[unit][ATA_ID_SECURITY_STATUS] & ATA_F_SEC_LOCKED)) {
        printf("not ");
    }
    printf("locked\n");

    /* printf("\tDrive is ");
    if (!ata_device_is_sce(unit)) {
        printf("not ");
    }
    printf("a Sony Playstation 2 HDD unit.\n"); */
}
#endif
